tools/print_tock_memory_usage.py

```diff
@@ -0,0 +1,422 @@
+#!/usr/bin/env python3
+
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#
+# Prints out the memory usage of a Tock kernel binary ELF.
+# Currently only works on ARM binaries.
+#
+# Usage: print_tock_memory_usage.py ELF
+#
+# Author: Philip Levis <philip.levis@gmail.com>
+
+# pylint: disable=superfluous-parens
+'''
+Script to print out the memory usage of a Tock kernel binary ELF.
+
+Usage: print_memory_usage.py ELF
+Options:
+  -dn, --depth=n      Group symbols at depth n or greater. E.g.,
+                      depth=2 will group all h1b::uart:: symbols
+                      together. Default: 1
+  -v, --verbose       Print verbose output.
+  -s, --show-waste    Show where RAM is wasted (due to padding)
+'''
+
+import os
+import re
+import sys
+import getopt
+import cxxfilt   # Demanging C++/Rust symbol names
+
+
+verbose = False
+show_waste = False
+symbol_depth = 1
```

Comment:
I don't get why these have to be global variables, is it required for `CLI parser`?

I would prefer if those were created in main and passed around when they are needed.

Reply:
There are a lot of variables -- not just these, but also the state of execution (e.g., the symbols). For a simple script like this I think it makes more sense to just have some globals: this isn't intended to be a reusable library. Otherwise we're passing around 5+ more variables to each function, or using tuples (which you just have to know which element is which), or dictionaries.

Reply:
I chose to not do this, as it would significantly re-architect the script and I'm not very aware of Python best practices for passing around lots of variables. I'm totally fine if someone else wants to do it after it's merged.

Reply:
Ok, it is fine for me thanks for the explanation.

I have not written python for a while but maybe it could make sense to split it out to separate classes where those variables could be encapsulated or rewrite the entire thing in Rust :)be